File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/homes/postal/HomePostalCoordinator.kt
package app.pantopus.android.ui.screens.homes.postal

import app.pantopus.android.data.api.models.homes.HomeResidencyAddressSnapshot
import app.pantopus.android.data.homes.HomePostalCodeRequest
import app.pantopus.android.data.homes.HomePostalCodec
import app.pantopus.android.data.homes.HomePostalKind
import app.pantopus.android.data.homes.HomePostalMailRequest
import app.pantopus.android.data.homes.HomePostalOutcome
import app.pantopus.android.data.homes.HomePostalScope
import app.pantopus.android.data.homes.PendingHomePostalCommand
import app.pantopus.android.data.homes.PendingHomePostalStore
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

enum class HomePostalAction { Submit, Check, Cancel }

fun interface HomePostalTransport {
    suspend fun resolve(
        draft: PendingHomePostalCommand,
        action: HomePostalAction,
    ): HomePostalOutcome
}

/** Retain one exact mailing or code attempt through restart and uncertain writes. */
class HomePostalCoordinator(
    val scope: HomePostalScope,
    private val store: PendingHomePostalStore,
    private val codec: HomePostalCodec,
    private val transport: HomePostalTransport,
    private val requireCurrent: suspend () -> Unit,
    private val newRequestId: () -> String = { UUID.randomUUID().toString() },
) {
    var pending: PendingHomePostalCommand? = null
        private set
    var isBusy = false
        private set
    var storageFailed = false
        private set
    private var revision = 0L
    private var expectedRequestId: String? = null
    private var knownOutcome: HomePostalOutcome? = null
    private var acknowledging: PendingHomePostalCommand? = null
    val outcome: HomePostalOutcome? get() = knownOutcome ?: pending?.outcome

    suspend fun restore() {
        val opening = revision
        requireCurrent()
        check(scope.isValid()) { CHANGED }
        val saved = readSaved()
        requireCurrent()
        check(opening == revision) { CHANGED }
        // IO cancellation can arrive after a confirmed acknowledgement removed
        // the slot. Reconcile that explicit local action before the old hint.
        val acknowledgement = acknowledging
        if (acknowledgement != null && (saved == null || !acknowledgement.sameIntent(saved))) clearAcknowledgedOriginal()
        check(expectedRequestId == null || saved?.requestId == expectedRequestId) { CHANGED }
        check(pending == null || saved == null || checkNotNull(pending).sameIntent(saved)) { CHANGED }
        val known = knownOutcome
        if (known != null) {
            check(saved != null && known.matches(saved)) { CHANGED }
            check(!known.isTerminal || saved.outcome?.isTerminal != true || known.sameDecision(saved.outcome)) { CHANGED }
        }
        pending = saved
        expectedRequestId = saved?.requestId
        if (known?.isTerminal != true) knownOutcome = saved?.outcome
    }

    suspend fun prepareMail(
        address: HomeResidencyAddressSnapshot,
        originalRequestId: String? = null,
    ) {
        val id = originalRequestId ?: newRequestId()
        prepare(PendingHomePostalCommand(scope, id, HomePostalKind.Mail, null, codec.encode(HomePostalMailRequest(id, address))))
    }

    suspend fun prepareCode(
        code: String,
        postcardId: String,
    ) {
        val id = newRequestId()
        prepare(PendingHomePostalCommand(scope, id, HomePostalKind.Code, postcardId, codec.encode(HomePostalCodeRequest(id, code))))
    }

    private suspend fun prepare(original: PendingHomePostalCommand) {
        val opening = revision
        requireCurrent()
        check(!isBusy && pending == null && knownOutcome == null && readSaved() == null) { CHANGED }
        check(opening == revision && codec.valid(original, scope)) { CHANGED }
        requireCurrent()
        replace(null, original)
        requireCurrent()
        check(opening == revision) { CHANGED }
        pending = original
        expectedRequestId = original.requestId
    }

    suspend fun resolve(action: HomePostalAction): HomePostalOutcome {
        requireCurrent()
        check(!isBusy && activeScopes.add(scope)) { "Your original postal request is still being checked. Try again shortly." }
        isBusy = true
        val opening = revision
        try {
            restore()
            check(opening == revision) { CHANGED }
            val draft = checkNotNull(pending) { CHANGED }
            val known = knownOutcome
            if (known?.isTerminal == true) {
                persist(known, draft)
                return known
            }
            val result = transport.resolve(draft, action)
            requireCurrent()
            check(opening == revision && result.matches(draft)) { CHANGED }
            persist(result, draft)
            return checkNotNull(outcome)
        } finally {
            isBusy = false
            activeScopes.remove(scope)
        }
    }

    suspend fun acknowledge(): PendingHomePostalCommand {
        val opening = revision
        requireCurrent()
        val original = checkNotNull(pending) { CHANGED }
        val proof = checkNotNull(outcome) { CHANGED }
        val saved = checkNotNull(readSaved()) { CHANGED }
        check(opening == revision && !isBusy && proof.isTerminal && original.sameIntent(saved)) { CHANGED }
        check(saved.outcome?.sameDecision(proof) == true) { CHANGED }
        requireCurrent()
        acknowledging = original
        replace(saved, null)
        clearAcknowledgedOriginal()
        requireCurrent()
        check(opening == revision) { CHANGED }
        return original
    }

    private fun clearAcknowledgedOriginal() {
        pending = null
        knownOutcome = null
        expectedRequestId = null
        acknowledging = null
    }

    fun hide() {
        revision++
        pending = null
        if (knownOutcome?.isTerminal != true) knownOutcome = null
    }

    private suspend fun persist(
        result: HomePostalOutcome,
        draft: PendingHomePostalCommand,
    ) {
        val opening = revision
        val saved = checkNotNull(readSaved()) { CHANGED }
        requireCurrent()
        check(opening == revision && saved.sameIntent(draft)) { CHANGED }
        val previous = knownOutcome?.takeIf { it.isTerminal } ?: saved.outcome
        val retained =
            if (previous?.isTerminal == true) {
                check(!result.isTerminal || previous.sameDecision(result)) { CHANGED }
                previous
            } else {
                result.projected()
            }
        knownOutcome = retained
        val confirmed = saved.copy(outcome = retained)
        pending = confirmed
        replace(saved, confirmed)
        requireCurrent()
        check(opening == revision) { CHANGED }
    }

    private suspend fun readSaved(): PendingHomePostalCommand? {
        try {
            return store.read(scope).also {
                check(it == null || codec.valid(it, scope)) { STORAGE }
                storageFailed = false
            }
        } catch (cancelled: kotlinx.coroutines.CancellationException) {
            throw cancelled
        } catch (_: Exception) {
            storageFailed = true
            error(STORAGE)
        }
    }

    private suspend fun replace(
        expected: PendingHomePostalCommand?,
        next: PendingHomePostalCommand?,
    ) {
        try {
            store.replace(scope, expected, next)
            storageFailed = false
        } catch (cancelled: kotlinx.coroutines.CancellationException) {
            throw cancelled
        } catch (_: Exception) {
            storageFailed = true
            error(STORAGE)
        }
    }

    companion object {
        private val activeScopes = ConcurrentHashMap.newKeySet<HomePostalScope>()
        const val CHANGED = "The saved postal request or your session changed. Reopen mail verification to recover the original."
        const val STORAGE = "The original postal request could not be read or saved. Keep it and retry recovery before starting another."
    }
}

File: frontend/apps/ios/Pantopus/Features/Homes/Postal/HomePostalCoordinator.swift
import Foundation
import Observation

/// One protected original postal command survives interruption and uncertain transport.
@Observable
@MainActor
final class HomePostalCoordinator {
    private(set) var pending: PendingHomePostalCommand?
    private(set) var isBusy = false
    private(set) var storageFailed = false
    private let scope: HomePostalScope
    private let store: any PendingHomePostalStoring
    private let transport: any HomePostalTransport
    private let requireCurrent: () throws -> Void
    private let requestId: () -> String
    private var revision = 0
    private var knownOutcome: HomePostalOutcome?
    private var expectedRequestId: String?
    private static var activeScopes = Set<HomePostalScope>()

    init(
        scope: HomePostalScope,
        store: any PendingHomePostalStoring,
        transport: any HomePostalTransport,
        requireCurrent: @escaping () throws -> Void,
        requestId: @escaping () -> String = { UUID().uuidString.lowercased() }
    ) {
        self.scope = scope
        self.store = store
        self.transport = transport
        self.requireCurrent = requireCurrent
        self.requestId = requestId
    }

    var outcome: HomePostalOutcome? {
        knownOutcome ?? pending?.outcome
    }

    func restore() throws {
        try requireCurrent()
        let saved = try readSaved()
        if let expectedRequestId, saved?.requestId != expectedRequestId { throw HomePostalError.changed }
        if let pending, let saved, !pending.hasSameIntent(as: saved) { throw HomePostalError.changed }
        if let knownOutcome {
            guard let saved, knownOutcome.matches(saved) else { throw HomePostalError.changed }
            if knownOutcome.isTerminal, let proof = saved.outcome, proof.isTerminal, !knownOutcome.hasSameDecision(as: proof) {
                throw HomePostalError.changed
            }
        }
        pending = saved
        expectedRequestId = saved?.requestId
        if knownOutcome?.isTerminal != true { knownOutcome = saved?.outcome }
    }

    func prepareMail(address: JSONValue, originalRequestId: String? = nil) throws {
        let id = originalRequestId ?? requestId()
        try prepare(PendingHomePostalCommand(
            scope: scope,
            requestId: id,
            kind: .mail,
            postcardId: nil,
            body: .object(["request_id": .string(id), "address": address])
        ))
    }

    func prepareCode(_ code: String, postcardId: String) throws {
        let id = requestId()
        try prepare(PendingHomePostalCommand(
            scope: scope,
            requestId: id,
            kind: .code,
            postcardId: postcardId,
            body: .object(["request_id": .string(id), "code": .string(code)])
        ))
    }

    private func prepare(_ draft: PendingHomePostalCommand) throws {
        try requireCurrent()
        guard !isBusy, pending == nil, knownOutcome == nil, try readSaved() == nil,
              draft.matches(scope) else { throw HomePostalError.changed }
        try replace(expected: nil, next: draft)
        pending = draft
        expectedRequestId = draft.requestId
    }

    @discardableResult
    func resolve(_ action: HomePostalAction) async throws -> HomePostalOutcome {
        try requireCurrent()
        let key = scope
        guard !isBusy, Self.activeScopes.insert(key).inserted else { throw HomePostalError.busy }
        isBusy = true
        defer { isBusy = false
            Self.activeScopes.remove(key)
        }
        let openingRevision = revision
        try restore()
        guard let draft = pending else { throw HomePostalError.changed }
        if let knownOutcome, knownOutcome.isTerminal {
            try persist(knownOutcome, for: draft)
            return knownOutcome
        }
        let result = try await transport.resolve(draft, action: action)
        try requireCurrent()
        guard revision == openingRevision, result.matches(draft) else { throw HomePostalError.sessionChanged }
        try persist(result, for: draft)
        return outcome ?? result
    }

    /// Called only after the user acknowledges a confirmed terminal result.
    /// Navigation then reads current residency access and mailing status.
    @discardableResult
    func acknowledge() throws -> PendingHomePostalCommand {
        try requireCurrent()
        guard !isBusy, let pending, let outcome, outcome.isTerminal, let saved = try readSaved(),
              pending.hasSameIntent(as: saved), saved.outcome?.hasSameDecision(as: outcome) == true else {
            throw HomePostalError.changed
        }
        try replace(expected: saved, next: nil)
        self.pending = nil
        knownOutcome = nil
        expectedRequestId = nil
        return pending
    }

    func hide() {
        revision += 1
        pending = nil
        // A known terminal identity survives a failed storage write in memory;
        // the protected original still permits recovery after process death.
        if knownOutcome?.isTerminal != true { knownOutcome = nil }
    }

    private func persist(_ result: HomePostalOutcome, for draft: PendingHomePostalCommand) throws {
        guard let saved = try readSaved(), saved.hasSameIntent(as: draft) else { throw HomePostalError.changed }
        let previous = knownOutcome?.isTerminal == true ? knownOutcome : saved.outcome
        let retained: HomePostalOutcome
        if let previous, previous.isTerminal {
            guard !result.isTerminal || previous.hasSameDecision(as: result) else { throw HomePostalError.changed }
            retained = previous
        } else {
            retained = result
        }
        knownOutcome = retained
        var confirmed = saved
        confirmed.outcome = retained
        pending = confirmed
        try replace(expected: saved, next: confirmed)
    }

    private func readSaved() throws -> PendingHomePostalCommand? {
        do {
            let saved = try store.load(scope: scope)
            guard saved == nil || saved?.matches(scope) == true else { throw HomePostalError.storage }
            storageFailed = false
            return saved
        } catch {
            storageFailed = true
            throw HomePostalError.storage
        }
    }

    private func replace(expected: PendingHomePostalCommand?, next: PendingHomePostalCommand?) throws {
        do {
            try store.replace(scope: scope, expected: expected, next: next)
            storageFailed = false
        } catch {
            storageFailed = true
            throw HomePostalError.storage
        }
    }
}
